include/foxy/detail/relay.hpp

```diff
@@ -127,115 +178,113 @@ relay_op<Stream, RelayHandler>::operator()(boost::system::error_code ec,
   auto& s = *p_;
   BOOST_ASIO_CORO_REENTER(*this)
   {
-    BOOST_ASIO_CORO_YIELD
-    s.server.async_read_header(s.req_parser, std::move(*this));
-    if (ec) { goto upcall; }
+    if (!s.req_parser.is_header_done()) {
+      BOOST_ASIO_CORO_YIELD
+      s.server.async_read_header(s.req_parser, std::move(*this));
+      if (ec) { goto upcall; }
+    }
 
     // remove hop-by-hop headers here and then store them externally...
-    // however, if the user is writing a Connection: close, they are
-    // communicating to our proxy that they wish to terminate the connection
-    // and we want to empower users to gracefully close their server sessions
-    // so we propagate the Connection: close field to convey to the remote
-    // that we won't be needing to persist the connection beyond this current
-    // response cycle
+    // however, if the user is writing a Connection: close, they are communicating to our proxy that
+    // they wish to terminate the connection and we want to empower users to gracefully close their
+    // server sessions so we propagate the Connection: close field to convey to the remote that we
+    // won't be needing to persist the connection beyond this current response cycle
     //
     BOOST_ASIO_CORO_YIELD
     {
-      auto& req = s.req_parser.get();
-
-      s.close_tunnel = !req.keep_alive();
-
-      auto const is_chunked = req.chunked();
+      s.close_tunnel        = s.close_tunnel || !s.req.keep_alive();
+      auto const is_chunked = s.req.chunked();
 
-      ::foxy::detail::export_connect_fields<fields_type>(req, s.req_fields);
+      ::foxy::detail::export_connect_fields<fields>(s.req, s.req_fields);
 
-      if (s.close_tunnel) { req.keep_alive(false); }
-      if (is_chunked) { req.chunked(true); }
+      if (s.close_tunnel) { s.req.keep_alive(false); }
+      if (is_chunked) { s.req.chunked(true); }
```

Comment:
This logic should be encapsulated in a free function which is independent of the stream, and operates only on the messages.